File: Web_Stadium/Services/SuspensionService.cs
using Microsoft.EntityFrameworkCore;
using Web_Stadium.EFCore;

namespace Web_Stadium.Services
{
    /// <summary>
    /// Xử lý treo giò cầu thủ sau mỗi trận
    /// Logic: Thẻ đỏ → treo N trận / Tích lũy thẻ vàng → treo 1 trận
    /// </summary>
    public class SuspensionService
    {
        private readonly SanBongContext _context;
        public SuspensionService(SanBongContext context) => _context = context;

        // ══════════════════════════════════════════════════════════
        // Gọi sau mỗi trận Closed — kiểm tra và áp dụng treo giò
        // ══════════════════════════════════════════════════════════
        public async Task XuLyTreoGio(int giaiDauId)
        {
            var giai = await _context.GiaiDaus
                .Include(g => g.DoiBongs).ThenInclude(d => d.ThanhViens)
                .FirstOrDefaultAsync(g => g.Id == giaiDauId);

            if (giai == null) return;

            // Lấy toàn bộ sự kiện thẻ của giải
            var suKiens = await _context.SuKienTrans
                .Include(s => s.TranDau)
                .Where(s => s.TranDau.GiaiDauId == giaiDauId
                         && s.ThanhVienId != null
                         && (s.LoaiSuKien == "TheVang"
                          || s.LoaiSuKien == "TheVangLan2"
                          || s.LoaiSuKien == "TheDo"))
                .ToListAsync();

            foreach (var tv in giai.DoiBongs.SelectMany(d => d.ThanhViens))
            {
                var suKienCuaTv = suKiens
                    .Where(s => s.ThanhVienId == tv.Id)
                    .ToList();

                // Treo giò vì thẻ đỏ trực tiếp
                var soTheDo = suKienCuaTv.Count(s => s.LoaiSuKien == "TheDo");

                // Treo giò vì tích lũy thẻ vàng
                var soTheVang = suKienCuaTv.Count(s =>
                    s.LoaiSuKien == "TheVang" || s.LoaiSuKien == "TheVangLan2");
                var treoDoVang = soTheVang / giai.SoTheVangTichLuy;

                var soTranTreo = soTheDo * giai.SoTranTreoGioTheDo + treoDoVang;

                // Cập nhật số trận còn bị cấm
                if (tv.SoTranTreoGio != soTranTreo)
                {
                    tv.SoTranTreoGio = soTranTreo;
                }
            }

            await _context.SaveChangesAsync();
        }

        // ══════════════════════════════════════════════════════════
        // Kiểm tra cầu thủ có bị treo giò ở trận cụ thể không
        // ══════════════════════════════════════════════════════════
        public bool BiTreoGio(ThanhVienDoi thanhVien, int soTranDaThiDau)
        {
            // Tính số trận đã thực sự thi đấu của cầu thủ
            // SoTranTreoGio = tổng số trận cần nghỉ tính từ đầu giải
            return thanhVien.SoTranTreoGio > soTranDaThiDau;
        }

        // ══════════════════════════════════════════════════════════
        // Lấy danh sách cầu thủ bị treo giò ở vòng kế tiếp
        // ══════════════════════════════════════════════════════════
        public async Task<List<ThanhVienDoi>> GetDanhSachTreoGio(
            int giaiDauId, int vongKeTiep)
        {
            var giai = await _context.GiaiDaus
                .Include(g => g.DoiBongs).ThenInclude(d => d.ThanhViens)
                .Include(g => g.TranDaus).ThenInclude(t => t.SuKiens)
                .FirstOrDefaultAsync(g => g.Id == giaiDauId);

            if (giai == null) return new();

            return giai.DoiBongs
                .SelectMany(d => d.ThanhViens)
                .Where(tv => tv.SoTranTreoGio >= vongKeTiep)
                .ToList();
        }
    }
}

File: tournament-service/src/main/java/com/pitchhub/tournament/service/SuspensionService.java
package com.pitchhub.tournament.service;

import com.pitchhub.tournament.model.GiaiDau;
import com.pitchhub.tournament.model.SuKienTran;
import com.pitchhub.tournament.model.ThanhVienDoi;
import com.pitchhub.tournament.repository.GiaiDauRepository;
import com.pitchhub.tournament.repository.SuKienTranRepository;
import com.pitchhub.tournament.repository.ThanhVienDoiRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Xu ly treo gio cau thu sau moi tran.
 * Logic: The do -> treo N tran / Tich luy the vang -> treo 1 tran.
 * Port truc tiep tu SuspensionService.cs — goi dong bo sau khi tran Closed
 * (KHONG dung background job dinh ky, da bo TournamentBackgroundService).
 */
@Service
public class SuspensionService {

    private final GiaiDauRepository giaiDauRepository;
    private final SuKienTranRepository suKienTranRepository;
    private final ThanhVienDoiRepository thanhVienDoiRepository;

    public SuspensionService(GiaiDauRepository giaiDauRepository,
                              SuKienTranRepository suKienTranRepository,
                              ThanhVienDoiRepository thanhVienDoiRepository) {
        this.giaiDauRepository = giaiDauRepository;
        this.suKienTranRepository = suKienTranRepository;
        this.thanhVienDoiRepository = thanhVienDoiRepository;
    }

    @Transactional
    public void xuLyTreoGio(Integer giaiDauId) {
        GiaiDau giai = giaiDauRepository.findDetailById(giaiDauId).orElse(null);
        if (giai == null) return;

        List<Integer> tranDauIds = giai.getTranDaus().stream().map(t -> t.getId()).toList();
        List<SuKienTran> suKiens = suKienTranRepository.findByTranDauIdIn(tranDauIds).stream()
                .filter(s -> s.getThanhVienId() != null
                        && ("TheVang".equals(s.getLoaiSuKien())
                        || "TheVangLan2".equals(s.getLoaiSuKien())
                        || "TheDo".equals(s.getLoaiSuKien())))
                .toList();

        for (var doi : giai.getDoiBongs()) {
            for (ThanhVienDoi tv : doi.getThanhViens()) {
                List<SuKienTran> suKienCuaTv = suKiens.stream()
                        .filter(s -> tv.getId().equals(s.getThanhVienId()))
                        .toList();

                long soTheDo = suKienCuaTv.stream().filter(s -> "TheDo".equals(s.getLoaiSuKien())).count();
                long soTheVang = suKienCuaTv.stream()
                        .filter(s -> "TheVang".equals(s.getLoaiSuKien()) || "TheVangLan2".equals(s.getLoaiSuKien()))
                        .count();
                long treoDoVang = soTheVang / giai.getSoTheVangTichLuy();

                int soTranTreo = (int) (soTheDo * giai.getSoTranTreoGioTheDo() + treoDoVang);

                if (!tv.getSoTranTreoGio().equals(soTranTreo)) {
                    tv.setSoTranTreoGio(soTranTreo);
                    thanhVienDoiRepository.save(tv);
                }
            }
        }
    }

    public boolean biTreoGio(ThanhVienDoi thanhVien, int soTranDaThiDau) {
        return thanhVien.getSoTranTreoGio() > soTranDaThiDau;
    }

    public List<ThanhVienDoi> getDanhSachTreoGio(Integer giaiDauId, int vongKeTiep) {
        GiaiDau giai = giaiDauRepository.findDetailById(giaiDauId).orElse(null);
        if (giai == null) return List.of();

        return giai.getDoiBongs().stream()
                .flatMap(d -> d.getThanhViens().stream())
                .filter(tv -> tv.getSoTranTreoGio() >= vongKeTiep)
                .toList();
    }
}
